RFC 9 datatypes descriptive variant (#61)
tango-controls/rfc: 9/README.md

  resources than a value of Data Type DevLong.

### DataType

* Any value of Properties, Attributes, Pipes' fields and Commands' input and output arguments SHALL have defined its 
  DataType.
  
* DataType SHALL be as follows: 

``` ABNF
 DataType = DevVoid 
 DataType =/ DevBoolean 
 DataType =/ DevFloat / DevDouble 
 DataType =/ DevShort / DevLong / DevLong64
 DataType =/ DevUChar / DevUShort / DevULong / DevULong64
 DataType =/ DevVarBooleanArray / DevVarDoubleArray / DevVarFloatArray
 DataType =/ DevString 
 DataType =/ DevVarShortArray / DevVarLongArray / DevVarLong64Array / DevVarCharArray
 DataType =/ DevVarStringArray
 DataType =/ DevVarUShortArray / DevVarULongArray / DevVarULong64Array
 
 DataType =/ DevEncoded / DevVarEncodedArray
 
 DataType =/ DevVarLongStringArray / DevVarDoubleStringArray
 
 DataType =/ DevState
  
``` 

### Numeric

#### DevBoolean

DevBoolean is a data type which values represents logical state of true or false.

* DevBoolean SHALL be defined as follows:

```ABNF
DevBoolean = "DEVBOOLEAN" | "DevBoolean"
```

* Values which DataType is DevBoolean SHALL be mapped to dev-boolean-value specification:

```abnf

> dev-boolen-value = BIT

Reviewer: Well, that is very dangerous assumption. In Java, for instance, **boolean** is 1 BYTE, not a BIT, in fact in C++ it is the same, see for instance on [SO](https://stackoverflow.com/questions/4626815/why-is-a-boolean-1-byte-and-not-1-bit-of-size). 

I would rather say "DevBoolean MUST be mappable to downstream language boolean spec"

Author: Yes, I will correct this.

* To display in the same way a numeric values read from a devices when some of them are using big-endian encoding and 
  other use little-endian encoding.

* To limit information processed to a practical level by choosing a Data Type of certain precision.
  For example, a value of Data Type DevShort typically provides less precise information and consumes less computing 
  resources than a value of Data Type DevLong.

### DataType

* Any value of Properties, Attributes, Pipes' fields and Commands' input and output arguments SHALL have defined its 
  DataType.
  
* DataType SHALL be as follows: 

```abnf
 DataType = DevVoid 

 DataType =/ DevBoolean 
 
 DataType =/ DevShort / DevLong / DevLong64
 DataType =/ DevUChar / DevUShort / DevULong / DevULong64
 
 DataType =/ DevFloat / DevDouble

 DataType =/ DevString  
 
 DataType =/ DevVarBooleanArray / DevVarDoubleArray / DevVarFloatArray
 DataType =/ DevVarShortArray / DevVarLongArray / DevVarLong64Array / DevVarCharArray
 DataType =/ DevVarStringArray
 DataType =/ DevVarUShortArray / DevVarULongArray / DevVarULong64Array
 
 DataType =/ DevEncoded / DevVarEncodedArray
 
 DataType =/ DevVarLongStringArray / DevVarDoubleStringArray
 

>  DataType =/ DevState

Reviewer 2: DevVarStateArray and DevEnum seem to be missing for the moment.
In tango_const.h from cppTango, I see as well the following types:
- ConstDevString
- DevInt
- DevPipeBlob

For the DevInt type, we need to clarify what it is and whether we really need it.

Author: Aren't ConstDevString and DevInt somehow C++ implementation specific? 
These are not defined in the tango.idl. Pogo, allows to use ConstDevSting for command but DevInt is not there at all.

Reviewer 2: > Aren't ConstDevString and DevInt somehow C++ implementation specific?
> These are not defined in the tango.idl.

Good point. So we should probably not mention these types.

Author: DevPipeBlob is there, now :-)

  
``` 

### Numeric

#### DevBoolean

DevBoolean is a data type which values represents logical state of true or false.

* DevBoolean SHALL be defined as follows:

```abnf

> DevBoolean = "DEVBOOLEAN" / "DevBoolean"

Reviewer 2: I don't know whether this is relevant but in cppTango, in enum CmdArgType, the enum associated to DevBoolean type is named DEV_BOOLEAN

Author: Thanks, @bourtemb. I will reserach this more, because it applies to other types, too.

Author: Together with @mliszcz,  we did a research and it seems there are no 'protocol' defined names for types except "CamelCase" strings commonly used in documentation and GUI applications. 

The `DEVBOOLEAN` and `DEV_BOOLEAN` are labels in enums for java and cpp, respectively. Even more, JTango is using also `Tango_DEV_BOOLEAN`. 

It seems, the more important are long values assigned to types, which can be found from both enums and consts defined in [JTango](https://github.com/tango-controls/JTango/blob/14a6527e3ca67339e51a905ef6c55d773ed04cb7/common/src/main/java/fr/esrf/TangoDs/TangoConst.java#L268) and probably related to the index of  [CmdArgTypeName array](https://github.com/tango-controls/cppTango/blob/4d2ad6555fd6bc6e6603406ea83b6ed0ed797eb3/cppapi/server/tango_const.h.in#L1024 ).

So, I suggest to specify data types names as camel case strings only and add info about associated long values. @hardion, @Ingvord, @t-b, I would be glad having your opinion, too :-)

Author: Finally, I left only "DevCamelCase" names.

License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later 
version. This Specification is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even 
the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details. You should have received a copy of the GNU General Public License along with this program; if not, 
see <http://www.gnu.org/licenses>.

This Specification is a [free and open standard](http://www.digistan.org/open-standard:definition) and is governed by 
the Digital Standards Organization's [Consensus-Oriented Specification System](http://www.digistan.org/spec:1/COSS).

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", 
and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).

## Tango Data Types Specification

Data Types are specifying format and limits of pieces of information exchanged between Tango Controls components.   

### Goals

> Data Types aims to standardise a way how different kind of information is entered, displayed and processed within

Reviewer 2: Data Types aims => Data Types specification aims or Data Types aim

> a Tango Controls systems irrespectively to its encoding in a transport protocol or computers' CPU architecture.

Reviewer 2: a Tango Controls systems => a Tango Controls system

> It also allows to distinguish different kind of numeric data, string data, enumerations or multi-values data. 

Reviewer 2: different kind => different kinds?

### Use Cases

There are many use cases for Data Types. Some of them are listed below:

* To specify expected data format for attributes, commands and pipes.

> * To display in the same way a numeric values read from a device when some of them are using big-endian encoding and 

Reviewer 2: a numeric values => numeric values
                   or    => a numeric value

> * To display in the same way a numeric values read from a devices when some of them are using big-endian encoding and 

Reviewer 3: ```suggestion
* To display in the same way a numeric values read from a device when some of them are using big-endian encoding and 
```

>   other use little-endian encoding.

Reviewer 2: other => others

  resources than a value of Data Type DevLong.

### DataType

* Any value of Properties, Attributes, Pipes' fields and Commands' input and output arguments SHALL have defined its 
  DataType.
  
* DataType SHALL be as follows: 

```abnf
 DataType = DevVoid 

 DataType =/ DevBoolean 
 
 DataType =/ DevShort / DevLong / DevLong64
 DataType =/ DevUChar / DevUShort / DevULong / DevULong64
 
 DataType =/ DevFloat / DevDouble

 DataType =/ DevString  
 
 DataType =/ DevVarBooleanArray / DevVarDoubleArray / DevVarFloatArray
 DataType =/ DevVarShortArray / DevVarLongArray / DevVarLong64Array / DevVarCharArray
 DataType =/ DevVarStringArray
 DataType =/ DevVarUShortArray / DevVarULongArray / DevVarULong64Array
 
 DataType =/ DevEncoded / DevVarEncodedArray
 
 DataType =/ DevVarLongStringArray / DevVarDoubleStringArray
 
 DataType =/ DevState / DevVarStateArray

 DataType =/ DevEnum
  
 DataType =/ DevPipeBlob
 
 DataTpe =/ DevFailed
``` 

### Numeric

#### DevBoolean

DevBoolean is a data type which values represents logical state of true or false.

* DevBoolean SHALL be defined as follows:

```abnf

> DevBoolean = "DevBoolean"

Reviewer 2: In this specific case, it is not obvious that this line defines the name.

Author: Solved by this [commit](https://github.com/tango-controls/rfc/pull/61/commits/cf2b7f87ab36e3ce680c36eba2c20120441dca4a)

```

* Values which DataType is DevBoolean SHALL carry one bit of information.

#### Integer data types

> DevShort, DevLong, DevLong64 represent singed integer values. DevUChar, DevUShort, DevULong and DevULong64 represent

Reviewer 2: singed => signed

unsigned values.

* Integer DataType SHALL have the following names:

```abnf
 DevShort = "DevShort"
 DevLong = "DevLong"
 DevLong64 = "DevLong64"
 DevUChar = "DevUChar"
 DevUShort = "DevUShort"
 DevULong = "DevULong"
 DevULong64 = "DevULong64"
```

> * Values of integer data types SHALL be in defined ranges and the information it cary SHALL have number of bits

Reviewer 2: it cary => they carry?

>   specified in the table below:

Reviewer 2: specified => as specified?

 DevEncoded = "DevEncoded"
```

* Value of \<DevEncoded\> SHALL be according to the \<encoded-value\> rule:
```abnf
 encoded-value = encoded_format encoded_data
                 ; elements order implied by the above rule MAY be ignored
```
;where \<encoded_format\> has type DevString and \<encoded_data\> has type DevVarCharArray. 

DevVarLongStringArray and DevVarDoubleStringArray values are structures to carry sequences of numbers 
(integer or double precision floating point, respectively) along with sequences of text/string data.

* Names DevVarLongStringArray and DevVarDoubleStringArray SHALL follow the following rules:
```abnf
 DevVarLongStringArray = "DevVarLongStringArray"
 DevVarDoubleStringArray = "DevVarDoubleStringArray"
```

* Values of DevVarLongStringArray and DevVarDoubleStringArray SHALL follow rules \<long-string-value\> 
and \<double-string-value\>, respectively:
```abnf
 long-string-value = lvalue svalue 
 double-string-value = dvalue svalue
                       ; elements order implied by the above rules MAY be ignored
```
;where \<lvalue\> has type DevVarLongArray, \<dvalue\> has type DevVarDoubleArray and \<svalue\> has type 
DevVarStringArray.

#### State

> DevState type standardize a way of describing state of Device. Its values are from predefined set of 14 names.   

Reviewer 2: standardize => standardizes

> DavState type standardize a way of describing state of Device. Its values are from predefined set of 14 names.   

Reviewer 3: Typo fix:
```suggestion
DevState type standardize...  
```

Author: thanks @gwen-soleil, fix is coming.

* DevState name SHALL be according to the following rule:
```abnf

>  DevState = "DevState

Reviewer 3: ```suggestion
 DevState = "DevState"
```

Author: fix is coming

```

* Value of DevState type SHALL follow the rule \<dev-state-value\>:
```abnf
 dev-state-value = "ALARM" / "INSERT" / "STANDBY" / "CLOSE" / "MOVING" / "UNKNOWN" / "DISABLE" / "OFF"

>  dev-state-value =/ "EXTRACT" / "ON" / "FAULT" / "OPEN" / "INIT" / "RUNNING"

Reviewer 2: This is describing only the names of the different states, not the values...
I find this a bit confusing.
In the IDL, DevState is declared as an enum.

Author: Hi @bourtemb, [this commit](https://github.com/tango-controls/rfc/pull/61/commits/c7db89be1dbaefbb920c464323883db1df6175e9) proposes the change. Please check if it is ok.

```

#### Enumeration

DevEnum data type allows to assign string Labels to DevShort values. It is valid only for Attributes. 

* DevEnum name SHALL follow the following rule:
```abnf
 DevEnum = "DevEnum"
```

* Label values of DevEnum SHALL follow the rule \<devenum-label\>:
```abnf
 devenum-label = 1*VCHAR
```  

* DevEnum labels SHALL be unique in the context of an Attribute.
* Each of DevEnum labels SHALL correspond to one and unique DevShort value. 
* Related DevShort values SHALL be consecutive and SHALL start with 0. 
* Values of DevEnum DataType SHALL be transported as values of DevShort.

> * For any of Attribute of DevEnum data type, list of Labels SHALL be available as `enum_labels` Attribute property.

Reviewer 3: ```suggestion
* For any of Attribute of DevEnum data type, list of Labels SHALL be available as `enum_labels` Attribute property (cf RFC4). 
```

> * * For any of Attribute of DevEnum data type, list of Labels SHALL be available as `enum_labels` Attribute property (see [4/Attribute](/4)).

Reviewer 2: 1 star too much here at the beginning of the line?
For any of Attribute of => For any Attribute of

* Tango Controls SHALL allow a Tango Client to retrieve labels associated with the DevShort value.

#### Pipe

DevPipeBlob is a Data Type to transfer data related to Pipes (see 7/Pipe).

* DevPipeBlob SHALL have the following name:
```abnf
 DevPipeBlob = "DevPipeBlob"
```

* A value of DevPipeBlob type SHALL follow the \<pipe-blob-value\> rule:

```abnf
 pipe-blob-value = name blob-data
                   ; elements order implied by the above rule MAY be ignored

 name = 1*VCHAR 

 blob-data = *blob-data-element
 
 blob-data-element = name value inner-blob inner-blob-name 

>                      ; elements order implied by the above rule MAY be ignored    

Reviewer 2: inner-blob-name is not described it seems

Author: Solved with [this commit](https://github.com/tango-controls/rfc/pull/61/commits/2528659c31865cbbd06df59f5ce867dc34efce0a)

 inner-blob = blob-data  
```

#### Exceptions 

> Results of failed operations (exceptions) within the Tango Controls are send as values/messages of DevFailed data type.

Reviewer 2: are send => are sent

* DevFailed type name is as follows:
```abnf
 DevFailed = "DevFailed"
```

* Any value of DevFiled type SHALL follow \<dev-failed-value\> rule:
```abnf
 dev-failed-value = 1*error 

 error = reason severity desc origin

>  reason = *VCHAR ; SHALL be the reason of the failur

Reviewer 3: ```suggestion
 reason = *VCHAR ; SHALL be the reason of the failure
```

>  reason = *VCHAR ; SHALL be the reason of the failure

Author: ```suggestion
 reason = *VCHAR ; SHOULD be a symbolic name, without whitespace, which points the reason for the failure and is standardized for some errors thrown by the API. Examples: "API_AttrOptProp", "API_AttrNotPolled","API_CmdNotPolled", "API_CommandNotFound"
```

Author: @bourtemb and @gwen, what do you think about this?

 severity = "WARN" | "ERR" | "PANIC"

>  desc = *VCHAR ; SHOULD describe the failur in more details, eg. call stack

Reviewer 3: ```suggestion
 desc = *VCHAR ; SHOULD describe the failure in more details
```
For me:
-  the call stack is in the "origin" field and not the "desc" field.
- desc is a message describing what is occuring in english.
- "reason" is string without spaces that is standardized for some errors thrown by the API. examples: "API_AttrOptProp", "API_AttrNotPolled","API_CmdNotPolled", "API_CommandNotFound", .... 
@bourtemb What do you think? Should we list all reasons of the API in the Tango RFC? I have found 39 of them in the Java code.

Reviewer 2: @gwen-soleil, I think it's good to standardize the reason field for the most common errors, and for this I think it's good to have a place where these errors are listed, but I don't think this RFC is the best location to list them all. Maybe we can add a comment as you did saying that the reason field is standardized for some errors thrown by the API, with some examples.
We would probably need another RFC (_Errors_ or something like that) to list all of these standard errors.

Author: I have checked the tango_const.h and there is about 130 definitions of API_... reasons. However, there is no mechanism for forcing usage of only these names for reasons.  Are these compatible between java and c++ implementations?

Author: I suggest changing SHALL to SHOULD, and skip listing all the possible reasons because if there will be a need to add a new one in the RFC will be invalidated and possibly not compatible between java and c++.

Reviewer 2: > I have checked the tango_const.h and there is about 130 definitions of API_... reasons. However, there is no mechanism for forcing usage of only these names for reasons. Are these compatible between java and c++ implementations?

I guess many of them are compatible because the Java API was developed after the C++ one and was mainly based on it at the beginning so I guess many errors are common in these 2 languages but I would not be surprised to see some differences.
This is why I find it interesting to have another RFC listing at least the standardized reason fields for the most common errors coming from a Tango implementation, with a description of these errors if it makes sense.
Then, 2 different Tango implementations could refer to this RFC and use consistent reason fields for similar errors.

Author: You have a point. I will add this RFC to the list of ones which are waiting to be added :)

>  desc = *VCHAR ; SHOULD describe the failur in more details, eg. call stack

Reviewer 2: The call stack is actually usually deduced from the origin fields of the **list** of error**s**.
For me, `origin `contains only 1 method name (The one from which the exception is thrown), this is not a call stack. But a call stack can be deducted from the origin fields of the list of errors.

>  reason = *VCHAR ; SHOULD be a symbolic name, without whitespace, which points the reason for the failure and is standardized for some errors thrown by the API. Examples: "API_AttrOptProp", "API_AttrNotPolled","API_CmdNotPolled", "API_CommandNotFound"

Author: ```suggestion
 reason = *VCHAR ; SHOULD be a symbolic name, without whitespace, which points the reason for the failure and is standardized 
                        ; for some errors thrown by the API. Examples: "API_AttrOptProp", "API_AttrNotPolled","API_CmdNotPolled", "API_CommandNotFound"
```

>  origin = *VCHAR ; SHALL identify the operation or the tango object which caused the failure  

Author: ```suggestion
 origin = *VCHAR ; SHALL identify the operation or the tango object which caused the failure, eg. function name
```

* Integer DataType SHALL have the following names:

```abnf
 DevShort = "DevShort"
 DevLong = "DevLong"
 DevLong64 = "DevLong64"
 DevUChar = "DevUChar"
 DevUShort = "DevUShort"
 DevULong = "DevULong"
 DevULong64 = "DevULong64"
```

* Values of integer data types SHALL be in defined ranges and the information they carry SHALL have number of bits
  as specified in the table below:

| DataType | Range | Bits of information |
-----------|-------|---------------------|
| DevShort | -(2^15) to (2^15)-1 |         16 |
| DevLong  | -(2^31) to (2^31)-1 | 32 |
| DevLong64  | -(2^63) to (2^63)-1 | 64 |
| DevUChar | 0 to 255 | 8 |
| DevUShort | 0 to (2^16)-1 | 16 |
| DevULong | 0 to (2^32)-1 | 32 |
| DevULong64 | 0 to (2^64)-1 | 64 |

#### Floating point data types

DevFloat and DevDouble represent floating point values.

* DevFloat and DevDouble SHALL use the following names:

```abnf
 DevFloat = "DevFloat"
 DevDouble = "DevDouble"
```

* Values of DevFloat type SHALL be of IEEE 754 single precision floating-point format and MAY be encoded in 32 bits.
* Values of DevDouble type SHALL be of IEEE 754 double precision floating-point format and MAY be encoded in 64 bits.

#### String data type

> DevString values represents strings (sequence of characters).

Reviewer 2: represents => represent

* DevEnum name SHALL follow the following rule:
```abnf
 DevEnum = "DevEnum"
```

* Label values of DevEnum SHALL follow the rule \<devenum-label\>:
```abnf
 devenum-label = 1*VCHAR
```  

* DevEnum labels SHALL be unique in the context of an Attribute.
* Each of DevEnum labels SHALL correspond to one and unique DevShort value. 
* Related DevShort values SHALL be consecutive and SHALL start with 0. 
* Values of DevEnum DataType SHALL be transported as values of DevShort.

* For any Attribute of DevEnum data type, list of Labels SHALL be available as `enum_labels` Attribute property (see [4/Attribute](/4)).

* Tango Controls SHALL allow a Tango Client to retrieve labels associated with the DevShort value.

#### Pipe

DevPipeBlob is a Data Type to transfer data related to Pipes (see 7/Pipe).

* DevPipeBlob SHALL have the following name:
```abnf
 DevPipeBlob = "DevPipeBlob"
```

* A value of DevPipeBlob type SHALL follow the \<pipe-blob-value\> rule:

```abnf
 pipe-blob-value = name blob-data
                   ; elements order implied by the above rule MAY be ignored

 name = 1*VCHAR 

 blob-data = *blob-data-element
 

>  blob-data-element = name value inner-blob inner-blob-name 

Reviewer 2: I think it is not clear what `value ` is here.  (Scalar or array of any basic Tango type)
I also think inner-blob-name does not make sense when the inner-blob is empty.

Author: @bourtemb, this [commit](https://github.com/tango-controls/rfc/pull/61/commits/5fb3d9dce38176bc6b4339f208959f8f8b7f6df3) adds the missing `value` specification.

Author: Regarding the `inner-blob-name`, even it has no meaning in case of inner-blob is empty, the IDL specifies it as a part of the [DevPipeDataElt structure](https://github.com/tango-controls/tango-idl/blob/1e5edb84d966814ad367f2674ac9a5658b6724ac/tango.idl#L555).

                     ; elements order implied by the above rule MAY be ignored    

 inner-blob = blob-data  

 inner-blob-name = name
```

#### Exceptions 

Results of failed operations (exceptions) within the Tango Controls are sent as values/messages of DevFailed data type.

* DevFailed type name is as follows:
```abnf
 DevFailed = "DevFailed"
```

> * Any value of DevFiled type SHALL follow \<dev-failed-value\> rule:

Reviewer 2: DevFiled => DevFailed

Author: solved with recent commit

```abnf
 devenum-label = 1*VCHAR
```  

* DevEnum labels SHALL be unique in the context of an Attribute.
* Each of DevEnum labels SHALL correspond to one and unique DevShort value. 
* Related DevShort values SHALL be consecutive and SHALL start with 0. 
* Values of DevEnum DataType SHALL be transported as values of DevShort.

* For any Attribute of DevEnum data type, list of Labels SHALL be available as `enum_labels` Attribute property (see [4/Attribute](/4)).

* Tango Controls SHALL allow a Tango Client to retrieve labels associated with the DevShort value.

#### Pipe

DevPipeBlob is a Data Type to transfer data related to Pipes (see 7/Pipe).

* DevPipeBlob SHALL have the following name:
```abnf
 DevPipeBlob = "DevPipeBlob"
```

* A value of DevPipeBlob type SHALL follow the \<pipe-blob-value\> rule:

```abnf
 pipe-blob-value = name blob-data
                   ; elements order implied by the above rule MAY be ignored

 name = 1*VCHAR 

 blob-data = *blob-data-element
 
 blob-data-element = name value inner-blob inner-blob-name 
                     ; elements order implied by the above rule MAY be ignored    

 inner-blob = blob-data  

>  inner-blob-name = name

Reviewer 2: I propose to remove this inner-blob and inner-blob-name to have something like:

`blob-data-element = name value / name blob-data`

Author: I suggest then `blob-data-element = name value / blob-data`

 inner-blob-name = name
```

Where \<value\> SHALL be a value of one of the following data types: DevVarBooleanArray, DevVarShortArray, DevVarLongArray,			
DevVarLong64Array, DevVarFloatArray, DevVarDoubleArray, DevVarCharArray, DevVarUShortArray,	DevVarULongArray,		

> DevVarULong64Array, DevVarStringArray, DevVarStateArray, DevState, DevVarEncodedArray, DevBoolean.

Reviewer 2: What about scalar DevDouble, DevFloat, ...?

Author: Apparently, according to IDL, there are no scalar data types. It is [AttrValUnion](https://github.com/tango-controls/tango-idl/blob/1e5edb84d966814ad367f2674ac9a5658b6724ac/tango.idl#L363).

Reviewer 2: According to the IDL. Yes, you're right.
I've seen the doc (https://tango-controls.readthedocs.io/en/latest/development/client-api/cpp-client-programmers-guide.html#reading-writing-device-pipe) is mentioning scalar data types too.
I think scalar are simply transported in arrays.

There is something confusing, I think.
It is the presence of DevBoolean type.
It is there only for the ATT_NO_DATA case.

Author: Yes, DevBoolean is only to poit there are no data :-)

Reviewer 2: [DevicePipeBlob::get_data_elt_type() method in cppTango](https://github.com/tango-controls/cppTango/blob/tango-9-lts/cppapi/client/devapi_pipe.cpp#L508-L517) distinguishes scalar and spectrum types.
So, from the user point of view, it seems like the data elements can be scalar Tango data types too.

Reviewer 2: And when looking deeper into cppTango code, it seems like the inner_blob_name is set to "scalar" string when the element is a scalar.

https://github.com/tango-controls/cppTango/blob/tango-9-lts/cppapi/client/devapi_pipe.cpp#L507-L513

and

https://github.com/tango-controls/cppTango/blob/tango-9-lts/cppapi/server/tango_const.h.in#L106-L107

Author: Yes, you are right. I think what matters here is users' perspectives, not a corba implementation. I will add the missing types.

Reviewer 2: It is interesting to note that the corresponding [JTango PipeDataElement::getType](https://github.com/tango-controls/JTango/blob/3c826dc478e1a4c2f3c6f2d154393184801f3e29/common/src/main/java/fr/esrf/TangoApi/PipeDataElement.java#L441) method does not seem to distinguish scalar and array types.

The PipeDataElement constructors seem to construct some DevPipeDataElt objects which are using a parameter which can be SCALAR or ARRAY, so this seems to be important at some point in Java too.

Author: I have added the [scalars](https://github.com/tango-controls/rfc/pull/61/commits/d804c9f03fa5a062695a9b359b763a0fc8884ce9)

Author: Should we create a list of potential pitfalls discovered during writing RFCs?

Reviewer 2: Why not? Where do you want to store them?

Author: I don know yet

 inner-blob-name = name
```

Where \<value\> SHALL be a value of one of the following data types: 

> DevVoid, DevBoolean, DevShort, DevLong, DevLong64, DevFloat, DevDouble, DevChar, DevUShort,	DevULong,		

Reviewer 2: I think in the case of pipe, DevVoid type does not make sense. I think it should be removed.
